### PR TITLE
core: import known blocks

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR fixes the known block insertion logic.

Geth node can import a batch of known blocks in such scenario:

* During the fast sync, the pivot point is already submitted but rollback
   happens. 
* Then node resets the head full block to a lower height via `rollback`
    and leaves a few known **full** blocks in the database.		
*  When node runs fast sync again, it can re-import a batch of known blocks via
    `insertChain` while a part of them have higher total difficulty than current
    head full block(new pivot point).

If the known blocks' total difficulty is lower than head full block, ignore them. But if the td is higher, commit them as the new head.

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -1178,13 +1178,23 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, []
...
+		// First block is still known block, the only scenario here is:
+		// We did a roll-back, and we want to re-import a batch of known blocks while a part
+		// of known blocks are higher than current head block.
+		if err == ErrKnownBlock {
+			block, err = bc.insertKnownChain(block, it)
+
+			if bc.CurrentBlock().NumberU64() != current {
+				lastCanon = bc.CurrentBlock()
+			}
+		}
```
[the remark this answers is not in the document]
Comment (holiman) [reply core/blockchain.go]: There might also be another quirk hidden here, which is not relevant for mainnet but might be for Clique... 
We check if `current >= block.NumberU64` to determine if the other block is 'higher' than ours. However, the _correct_ would be to check if the other difficulty is higher. The `insertKnownBlock` checks the correct metric, but the number-check might do it erroneously. A number-longer (but td-lower) sidechain might be known to us already, and would trigger this path. I'm not sure what would happen, but it should be fairly simple to make a testcase for it

File: core/blockchain.go [hunk 2]
```diff
@@ -1347,13 +1357,44 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, []
...
+		// Short circuit if the known block cannot be imported as a new
+		// canonical block.
+		if block.ParentHash() != current.Hash() || externTd.Cmp(localTd) <= 0 {
+			break
+		}
+		knownBlocks = append(knownBlocks, block)
+		localTd, current = new(big.Int).Add(localTd, block.Difficulty()), block
+	}
+	if len(knownBlocks) > 0 {
+		bc.insert(knownBlocks[len(knownBlocks)-1])
```
Comment (holiman) [inline core/blockchain.go]: I don't understand this. This inserts only the last block?

File: core/blockchain_test.go [hunk 3]
```diff
@@ -1810,3 +1699,230 @@ func TestPrunedImportSide(t *testing.T) {
 	testSideImport(t, 1, 10)
 	testSideImport(t, 1, -10)
 }
+
+func TestInsertKnownHeaders(t *testing.T)      { testInsertKnownChainData(t, "headers") }
```
Comment (holiman) [inline core/blockchain_test.go]: Great that you added testcases! Could we also have the following:

1. Two competing chains, where `A` is longer, but `B` has higher td, 
2. Both are imported (known). 
3. We feed chain `A` again into the importer to see what happens. Maybe in combination with a setHead where we rollback to some common ancestor

File: core/blockchain.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rjl493456442) [reply core/blockchain.go]: These are out dated.

File: core/blockchain_test.go (hunk 3 above)
Comment (rjl493456442) [reply core/blockchain_test.go]: Added

File: core/blockchain.go [hunk 4]
```diff
@@ -932,6 +932,26 @@ func (bc *BlockChain) WriteBlockWithoutState(block *types.Block, td *big.Int) (e
...
+
+	current := bc.CurrentBlock()
+	if block.ParentHash() != current.Hash() {
+		if err := bc.reorg(current, block); err != nil {
+			return err
+		}
+	}
+	// Write the positional metadata for transaction/receipt lookups.
+	// Preimages here is empty, ignore it.
+	rawdb.WriteTxLookupEntries(bc.db, block)
```
Comment (holiman) [inline core/blockchain.go]: Is not txlookupEntries written already within `reorg` ?

Comment (rjl493456442) [reply core/blockchain.go]: `reorg` doesn't write txLookUp for the new head, only for the reorged blocks.

Comment (rjl493456442) [reply core/blockchain.go]: Yes you are right. reorg also writes for head.

**Comment by rjl493456442** [comment]
@holiman I fix the `reorg` function. In the original implementation, `reorg` function will update the `canonical_hash` and write `txLookup` for all *new blocks include the new head*. But outside of this function, we will do the same thing for new head again. So basically in the `reorg` we only need to care about the *rebirth blocks* except the new head. Does it make sense to you?

There is one ugly thing in this implementation is `EventLogs`. In the normal case, the `reorg` function will collect all *rebirth logs* and broadcast them to subscribers. And outside of `reorg` function, we will collect the logs from new head block.

In theory, I should do the same thing for new head block which is **known block**. But I didn't. The reason is the `known block insertion` basically caused by syncing rollback. But in the rollback function, we don't notify subscribers these logs are **removed**. So if we fire the logs again in the `known block insertion`, users can receive two notifications.
But I think it's fine since geth node hasn't finished syncing that time. Probably we don't need to process the ugly log event logic. What's your opinion about it? cc@karalabe

File: core/blockchain.go [hunk 5]
```diff
@@ -1139,18 +1159,45 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, []
...
+		)
+		externTd = new(big.Int).Add(externTd, block.Difficulty())
+		for block != nil && err == ErrKnownBlock && localTd.Cmp(externTd) >= 0 {
 			stats.ignored++
 			block, err = it.next()
+
+			if block != nil {
+				externTd = new(big.Int).Add(externTd, block.Difficulty())
+			}
+		}
```
Comment (karalabe) [inline core/blockchain.go]: I think you can simplify this code segment:

```go
for block != nil && err == ErrKnownBlock {
	externTd = new(big.Int).Add(externTd, block.Difficulty())
	if localTd.Cmp(externTd) < 0 {
		break;
	}
	stats.ignored++
	block, err = it.next()
}
```
```

```

Comment (karalabe) [reply core/blockchain.go]: i.e. you don't need the same externTd += block.Diff calculation both outside and inside the for loop. One seems enough to me. Maybe I'm wrong.

File: core/blockchain.go [hunk 6]
```diff
@@ -1139,18 +1159,45 @@ func (bc *BlockChain) insertChain(chain types.Blocks, verifySeals bool) (int, []
...
+		// The remaining blocks are still known blocks, the only scenario here is:
+		// During the fast sync, the pivot point is already submitted but rollback
+		// happens. Then node resets the head full block to a lower height via `rollback`
+		// and leaves a few known blocks in the database.
+		//
+		// When node runs a fast sync again, it can re-import a batch of known blocks via
+		// `insertChain` while a part of them have higher total difficulty than current
+		// head full block(new pivot point).
+		if err == ErrKnownBlock {
```
Comment (karalabe) [inline core/blockchain.go]: I don't think you need this `err` check here. It's essentially the same as the for loop below. You can just get rid of the `if` and have the same code.

File: core/blockchain.go [hunk 7]
```diff
@@ -1524,15 +1571,15 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
 	} else {
 		log.Error("Impossible reorg, please file an issue", "oldnum", oldBlock.Number(), "oldhash", oldBlock.Hash(), "newnum", newBlock.Number(), "newhash", newBlock.Hash())
 	}
-	// Insert the new chain, taking care of the proper incremental order
-	for i := len(newChain) - 1; i >= 0; i-- {
+	// Insert the new chain(except the head block(reverse order)),
+	// taking care of the proper incremental order.
+	for i := len(newChain) - 1; i >= 1; i-- {
```
Comment (karalabe) [inline core/blockchain.go]: Why aren't we writing `newChain[0]`?

**Comment by karalabe** [comment]
PR *seems* good to me. Please fix the 2 nits + linter.

File: core/blockchain.go (hunk 7 above)
Comment (karalabe) [reply core/blockchain.go]: Ah, if I understand correctly, we're calling reorg from `writeBlockWithState`, so the injection is done outside already.

Comment (rjl493456442) [reply core/blockchain.go]: Exactly.

**Pull request merged**